File: stack/largest_rectangle/largest_rectangle.py

def find_largest_rectangle(heights: list[int]) -> int:  # LeetCode Q.84.
    prev_smaller_indices: list[int] = [-1] * len(heights)
    next_smaller_indices: list[int] = [len(heights)] * len(heights)

    prev_stack, next_stack = [], []  # Format: (height, idx).
    for idx, height in enumerate(heights):
        while prev_stack and prev_stack[-1][0] >= height:
            prev_stack.pop(-1)

        if prev_stack:
            past_idx = prev_stack[-1][1]
            prev_smaller_indices[idx] = past_idx

        prev_stack.append((height, idx))

        while next_stack and next_stack[-1][0] > height:
            past_idx = next_stack.pop(-1)[1]
            next_smaller_indices[past_idx] = idx

        next_stack.append((height, idx))

    largest_rectangle = 0
    for idx, height in enumerate(heights):
        width = 1
        width += next_smaller_indices[idx] - 1 - idx
        width += idx - 1 - prev_smaller_indices[idx]

        rectangle = height * width
        if rectangle > largest_rectangle:
            largest_rectangle = rectangle

    return largest_rectangle

File: stack/largest_rectangle/largest_rectangle.cpp
#include <vector>
#include <stack>
using namespace std;

int find_largest_rectangle(vector<int> &heights) // LeetCode Q.84.
{
    vector<int> prevSmallerIndices(heights.size(), -1);

    int maxArea = 0;

    stack<pair<int, int>> nextStack, prevStack; // Format: {idx, height}.

    for (int idx = 0; idx < heights.size(); idx++)
    {
        int height = heights[idx];

        // Prev smaller search.
        while (!prevStack.empty() && prevStack.top().second >= height)
            prevStack.pop();

        if (!prevStack.empty())
            prevSmallerIndices[idx] = prevStack.top().first;

        prevStack.push({idx, height});

        // Next smaller search.
        while (!nextStack.empty() && nextStack.top().second > height)
        {
            int prevIdx = nextStack.top().first;
            nextStack.pop();

            int width = idx - prevSmallerIndices[prevIdx] - 1;
            int area = heights[prevIdx] * width;
            if (area > maxArea)
                maxArea = area;
        }

        nextStack.push({idx, height});
    }

    while (!nextStack.empty())
    {
        int idx = nextStack.top().first;
        nextStack.pop();

        int width = heights.size() - prevSmallerIndices[idx] - 1;
        int area = heights[idx] * width;
        if (area > maxArea)
            maxArea = area;
    }

    return maxArea;
}
